keps/sig-node/2358-node-affinity-manager/README.md

```diff
@@ -0,0 +1,532 @@
+
+## Motivation
+
+In Kubernetes, "Node Affinity" controls how pods are scheduled. 
+However, there is currently no way to guarantee that the node affinity will still be met during runtime.
+
+### Goals
+
+- Uncomment the NodeAffinity `RequiredDuringSchedulingRequiredDuringExecution` field and corresponding validations.
+- Add a node affinity manager to the node lifecycle controllers to ensure pods being evicted if the selectors are no longer met at runtime.
+
+### Non-Goals
+
+## Proposal
+
+### User Stories
+
+#### Manage Nodes With Special Resources
+
+Assuming there are multiple nodes labeled with "app=cache", which means that cache apps like Redis can be scheduled on this node.
+I added some nodes to provide more available memory, and I wish to migrate all previous cache applications to these new nodes.
+With runtime NodeSelector constraints, I can simply delete the label from existing nodes to evict all cache apps.
+
+#### Manage Multi Tenant Cluster
+
+I have some nodes labeled with "userA=allow" and "userB=allow", which means that these nodes are available to user A and user B.
+As the scale grows, I added more nodes for user B, and all existing nodes are only available to user A.
+I can simply delete the "userB=allow" label from the existing nodes to re-schedule all pods of user B to these new nodes.
+
+#### Share Resources More Effectively
+
+Assuming I have some failure tolerance calculation apps that will continuously generate large temporary files, 
+with a `RequiredDuringSchedulingIgnoredDuringExecution` selector like "available-disk-gb>100" as the maximum constraint,
+and a `RequiredDuringSchedulingRequiredDuringExecution` selector like "available-disk-gb>20" as the minimal constraint.
+Since these large files are temporary, a 1000GB disk may support more than 10 applications that require a 100GB disk at the same time.
+I have a monitor service to label the available disk of a node periodically, such as "available-disk-gb=100".
+
+If a node occasionally assigned too many of this type apps before the disk label being updated by the monitor service, the node may hit the disk pressure dut to the overcommitment.
+Suppose the monitor service updates the label to "available-disk-gb=15" after a period of time,
+with runtime constraints, the pods with "available-disk-gb>20" will be evicted, but pods with "available-disk-gb>10" can survive.
+
+### Implementation Details/Notes/Constraints/Caveats
+
+Because the NodeAffinity only supports the node name and labels right now, and this field is immutable at runtime,
+the node affinity manager only listens to node label changes, and evict pods that have `RequiredDuringSchedulingRequiredDuringExecution` selectors but no longer met.
+It simply deletes the pods and expects the controller (e.g. DeploymentController) to re-create a pod, then the scheduler will take the scheduling task again.
+
+That means a pod without a controller will be deleted permanently.
```

Comment:
I filed a issue to discuss this behavior https://github.com/kubernetes/kubernetes/issues/97832

another issue related: https://github.com/kubernetes/kubernetes/issues/96363